File: Kotlin/src/main/kotlin/intermediate/boj_10986.kt
package intermediate

fun boj_10986() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine()
        .split(" ")
        .map { it.toInt() }

    val preSum = LongArray(n + 1) { 0 }
    val cnt = LongArray(m) { 0 }
    val arr = readLine()
        .split(" ")
        .map { it.toLong() }

    for (i in 1 until preSum.size) {
        preSum[i] = (preSum[i - 1] + arr[i - 1])
        cnt[(preSum[i] % m).toInt()]++
    }

    println("[ preSum ]")
    for (k in preSum)
        print("$k ")
    println()

    println("[ cnt ]")
    for (k in cnt)
        print("$k ")
    println()

    var ans = cnt[0]
    for (k in cnt)
        if (k >= 2)
            ans += k * (k - 1) / 2

    println(ans)
}

File: Cpp/Intermediate/boj_10986.cpp
#include <iostream>
#include <vector>

#define endl "\n"

using namespace std;

constexpr int MAX = 1e6 + 1;

vector<unsigned long long> preSum(MAX, 0);
vector<unsigned long long> cnt(MAX, 0);

void boj_10986() {
    int n, m;
    cin >> n >> m;

    for (int i = 1; i <= n; i++) {
        unsigned long long x;
        cin >> x;

        preSum[i] = (preSum[i - 1] + x) % m;
        cnt[preSum[i]]++;
    }

    unsigned long long ans = cnt[0];
    for (int i = 0; i < m; i++)
        if (cnt[i] >= 2)
            ans += (cnt[i] * (cnt[i] - 1)) / 2;

    cout << ans << endl;
}